NumLib/ODESolver/TimeLoopSingleODE.h

```diff
@@ -36,13 +36,18 @@ class TimeLoopSingleODE final
      * \param linear_solver the linear solver used to solve the linearized ODE
      *                      system.
      * \param nonlinear_solver The solver to be used to resolve nonlinearities.
+     * \param convergence_criterion The convergence criterion used by the
+     * nonlinear solver.
      */
-    TimeLoopSingleODE(TDiscODESys& ode_sys,
-                               std::unique_ptr<GlobalLinearSolver>&& linear_solver,
-                               std::unique_ptr<NLSolver>&& nonlinear_solver)
+    TimeLoopSingleODE(
+        TDiscODESys& ode_sys,
+        std::unique_ptr<GlobalLinearSolver>&& linear_solver,
+        std::unique_ptr<NLSolver>&& nonlinear_solver,
+        std::unique_ptr<ConvergenceCriterion>&& convergence_criterion)
```

Comment:
why not create a nonlinear solver object for each process? is there any reason to reuse it?